query-algebrizer/src/clauses/not.rs

```diff
@@ -56,3 +55,146 @@ impl ConjoiningClauses {
+            value_type: ValueType::String,
+            multival: true,
+            ..Default::default()
+        });
+        add_attribute(&mut schema, 68, Attribute {
+            value_type: ValueType::Long,
+            multival: false,
+            ..Default::default()
+        });
+        add_attribute(&mut schema, 69, Attribute {
+            value_type: ValueType::Long,
+            multival: false,
+            ..Default::default()
+        });
+        schema
+    }
+
+    // Simple alternation.
+    #[test]
+    fn test_not() {
+        let schema = prepopulated_schema();
+        let query = r#"
+            [:find ?x
+             :where [?x :foo/knows "John"]
+                    (not [?x :foo/parent "Ámbar"]
+                         [?x :foo/knows "Daphne"])]"#;
+        let cc = alg(&schema, query);
+
+        let vx = Variable::from_valid_name("?x");
+        let d0 = "datoms00".to_string();
+        let d1 = "datoms01".to_string();
+        let d2 = "datoms02".to_string();
+        let d0e = QualifiedAlias::new(d0.clone(), DatomsColumn::Entity);
+        let d0a = QualifiedAlias::new(d0.clone(), DatomsColumn::Attribute);
+        let d0v = QualifiedAlias::new(d0.clone(), DatomsColumn::Value);
+        let d1e = QualifiedAlias::new(d1.clone(), DatomsColumn::Entity);
+        let d1a = QualifiedAlias::new(d1.clone(), DatomsColumn::Attribute);
+        let d1v = QualifiedAlias::new(d1.clone(), DatomsColumn::Value);
+        let d2e = QualifiedAlias::new(d2.clone(), DatomsColumn::Entity);
+        let d2a = QualifiedAlias::new(d2.clone(), DatomsColumn::Attribute);
+        let d2v = QualifiedAlias::new(d2.clone(), DatomsColumn::Value);
+
+        let knows = QueryValue::Entid(66);
+        let subquery_knows = QueryValue::Entid(66);
+        let parent = QueryValue::Entid(67);
+        let john = QueryValue::TypedValue(TypedValue::typed_string("John"));
+        let ambar = QueryValue::TypedValue(TypedValue::typed_string("Ámbar"));
+        let daphne = QueryValue::TypedValue(TypedValue::typed_string("Daphne"));
+
+        let mut subquery = ConjoiningClauses::default();
+        subquery.from.append(&mut vec![SourceAlias(DatomsTable::Datoms, d1),
+                                  SourceAlias(DatomsTable::Datoms, d2)]);
+        subquery.column_bindings.insert(vx.clone(), vec![d1e.clone(), d2e.clone()]);
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1a.clone(), parent)));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1v.clone(), ambar)));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d2a.clone(), subquery_knows)));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d2v.clone(), daphne)));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1e.clone(), QueryValue::Column(d2e.clone()))));
```

Comment:
Missing here is the link between the `NOT EXISTS` and the body of the enclosing query. The CC you've built here actually encodes this query:

```
[:find ?x :where [?x :foo/name "John"] (not [?y :foo/parent "Ámbar"] [?y :foo/knows "Daphne"])]
```

— note that it's `?y` inside, not `?x`! — which is "find people named John unless there's a person who's Ámbar's child and knows Daphne".

I'd be expecting to see `ColumnConstraint::Equals(d0e.clone(), QueryValue::Column(d1e.clone())` here, and for the `column_bindings` to be pre-populated from the parent with mention of `d0`.